server_development/topics/vault.adoc

```diff
@@ -0,0 +1,44 @@
+[[_vault-spi]]
+== Vault SPI
+
+=== Vault provider
+
+You can use a vault SPI from `org.keycloak.vault` package to write custom extension for {project_name} to connect to arbitrary vault implementation.
+
+The built-in `plaintext` provider is an example of the implementation of this SPI. In general the following rules apply:
+
+* The vault provider should respect realm, i.e. an expression `${vault.secret-id}` would evaluate generally to different values of a secret `secret-id`, depending on whether it was used in a realm _A_ or realm _B_. This is to prevent secret leaking across realms. To differentiate between realms, the realm needs to be passed to the created `VaultProvider` instance from `VaultProviderFactory.create()` method where it is available from its the `KeycloakSession` parameter.
```

Comment:
I would rewrite this to something along the lines of "If you want to isolate or limit what secrets can be retrieved by a realm your provider should take into account the realm name when looking up secrets, for example by prefixing entries with the realm name.

Reply:
I like this one better, thanks!